core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,705 @@
+
+        Some(schedulable_threads)
+    }
+
+    /// Returns `ThreadSet` of schedulable threads for the given readable account.
+    /// If the account is not locked, then all threads are schedulable.
+    /// If only read locked, then all threads are schedulable.
+    /// If write-locked, then only the thread holding the write lock is schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn read_schedulable_threads(&self, account: &Pubkey) -> ThreadSet {
+        // If the account is only read locked, then a read lock can be taken on any thread
+        // that is not at the sequential limit.
+        self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
+            let mut schedulable_threads = ThreadSet::any(self.num_threads);
+            for thread_id in thread_set.contained_threads_iter() {
+                if counts[thread_id] == self.sequential_queue_limit {
+                    schedulable_threads.remove(thread_id);
+                }
+            }
+            schedulable_threads
+        })
+    }
+
+    /// Returns `ThreadSet` of schedulable threads for the given writable account.
+    /// If the account is not locked, then all threads are schedulable.
+    /// If read-locked on a single thread, then only that thread is schedulable.
+    /// If write-locked, then only that thread is schedulable.
+    /// In all other cases, no threads are schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn write_schedulable_threads(&self, account: &Pubkey) -> ThreadSet {
+        // If the account is only read locked, then a write lock can only be taken
+        // if the read lock is held by a single thread, and the limit is not exceeded.
+        self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
+            thread_set
+                .only_one_contained()
+                .filter(|thread_id| counts[*thread_id] < self.sequential_queue_limit)
+                .map(ThreadSet::only)
+                .unwrap_or_else(ThreadSet::none)
+        })
+    }
+
+    /// Returns `ThreadSet` of schedulable threads, given the read-only lock handler.
+    /// Helper function, since the only difference between read and write schedulable threads
+    /// is in how the case where only read locks are held is handled.
+    /// If there are no locks, then all threads are schedulable.
+    /// If only write-locked, then only the thread holding the write lock is schedulable.
+    /// If a mix of locks, then only the write thread is schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn schedulable_threads_with_read_only_handler(
+        &self,
+        account: &Pubkey,
+        read_only_handler: impl Fn(&ThreadSet, &[u32]) -> ThreadSet,
```

Comment:
nit: `impl FnOnce` communicates better to the reader: https://doc.rust-lang.org/stable/std/ops/trait.FnOnce.html:

> Since both [Fn](https://doc.rust-lang.org/stable/std/ops/trait.Fn.html) and [FnMut](https://doc.rust-lang.org/stable/std/ops/trait.FnMut.html) are subtraits of FnOnce, any instance of [Fn](https://doc.rust-lang.org/stable/std/ops/trait.Fn.html) or [FnMut](https://doc.rust-lang.org/stable/std/ops/trait.FnMut.html) can be used where a FnOnce is expected.

> Use FnOnce as a bound when you want to accept a parameter of function-like type and only need to call it once. If you need to call the parameter repeatedly, use [FnMut](https://doc.rust-lang.org/stable/std/ops/trait.FnMut.html) as a bound; if you also need it to not mutate state, use [Fn](https://doc.rust-lang.org/stable/std/ops/trait.Fn.html).

Reply:
hmm, I tried to train myself by reading this pr numerous times. that said, still i don't think using closure is the best way. firstly, these apis are internal, so the design space is very flexible.

so, change this  fn sig. to `fn schedulable_threads(..., write_attempted: bool)` and move caller's closures' bodies in this single place for readability.

Reply:
That seems reasonable - though I modifed slightly in that I used a const generic `const WRITE: bool` parameter.